### ISSUE TITLE
libyang2: segmentation fault due to long module name

### ISSUE THREAD

**Issue by jvijtiuk** [CESNET/libyang#822]
Hi,

this file crashes yanglint built on the libyang2 branch.

```
module s0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000{yang-version 1.0;namespace 0;prefix d;container c{}grouping g0000{leaf l0000{type n{range 0{e:a;}}}}}
```

Regards,
Juraj

**Comment by rkrejci** [comment]
I'm not able to reproduce the issue, even valgrind does not report anything:
```
valgrind ./yanglint s.yang 
==28357== Memcheck, a memory error detector
==28357== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==28357== Using Valgrind-3.13.0 and LibVEX; rerun with -h for copyright info
==28357== Command: ./yanglint s.yang
==28357== 
err : Referenced type "n" not found. (/s0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000:{grouping='g0000'}/l0000)
==28357== 
==28357== HEAP SUMMARY:
==28357==     in use at exit: 0 bytes in 0 blocks
==28357==   total heap usage: 2,015 allocs, 2,015 frees, 800,921 bytes allocated
==28357== 
==28357== All heap blocks were freed -- no leaks are possible
==28357== 
==28357== For counts of detected and suppressed errors, rerun with: -v
==28357== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```
I'm testing on x86_64 (openSUSE), what is your environment?

**Comment by jvijtiuk** [comment]
I'm testing on x86_64 on Debian 10.

Here is the output when i run yanglint with valgrind, with the crash copied from this issue on github:

```
valgrind libyang/build/yanglint ~/crash.yang 
==10268== Memcheck, a memory error detector
==10268== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==10268== Using Valgrind-3.14.0 and LibVEX; rerun with -h for copyright info
==10268== Command: libyang/build/yanglint /home/user/crash.yang
==10268== 
err : Referenced type "n" not found. (/s0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000:{grouping='g0000'}/l0000)
==10268== Invalid read of size 1
==10268==    at 0x48675AC: lys_parse_fd_ (tree_schema.c:857)
==10268==    by 0x4867ADB: lys_parse_fd_module (tree_schema.c:878)
==10268==    by 0x4867ADB: lys_parse_path_ (tree_schema.c:910)
==10268==    by 0x40514D: main_ni (main_ni.c:711)
==10268==    by 0x403714: main (main.c:43)
==10268==  Address 0x768f058d482ef9 is not stack'd, malloc'd or (recently) free'd
==10268== 
==10268== 
==10268== Process terminating with default action of signal 11 (SIGSEGV)
==10268==  General Protection Fault
==10268==    at 0x48675AC: lys_parse_fd_ (tree_schema.c:857)
==10268==    by 0x4867ADB: lys_parse_fd_module (tree_schema.c:878)
==10268==    by 0x4867ADB: lys_parse_path_ (tree_schema.c:910)
==10268==    by 0x40514D: main_ni (main_ni.c:711)
==10268==    by 0x403714: main (main.c:43)
==10268== 
==10268== HEAP SUMMARY:
==10268==     in use at exit: 73,829 bytes in 273 blocks
==10268==   total heap usage: 1,983 allocs, 1,710 frees, 784,929 bytes allocated
==10268== 
==10268== LEAK SUMMARY:
==10268==    definitely lost: 2,515 bytes in 31 blocks
==10268==    indirectly lost: 484 bytes in 12 blocks
==10268==      possibly lost: 6,136 bytes in 12 blocks
==10268==    still reachable: 64,694 bytes in 218 blocks
==10268==         suppressed: 0 bytes in 0 blocks
==10268== Rerun with --leak-check=full to see details of leaked memory
==10268== 
==10268== For counts of detected and suppressed errors, rerun with: -v
==10268== ERROR SUMMARY: 2 errors from 1 contexts (suppressed: 0 from 0)
Segmentation fault
```

**Comment by rkrejci** [comment]
I'm still not able to reproduce even on Debian 10 (x86_64) in Vagrant. Some points:
* how you compile libyang (what are the options for `cmake`)?
* which compiler are you using?
* can you share the core dump file?

**Comment by jvijtiuk** [comment]
I have figured out why the issue is not reproducible. It only appears when compiled with the modified version of the compiler that the fuzzer uses to insert instrumentation, so I assume this might not be a bug at all. I guess will then have to start testing all the crashes with a version of libyang built with normal compilers from now on.

I have compiled libyang with these options
`cmake -DCMAKE_C_COMPILER=afl-clang-fast -DCMAKE_BUILD_TYPE=Release ..`

The compiler version looks like this:
`afl-clang-fast 2.52b by <lszekeres@google.com>
clang version 6.0.1-10 (tags/RELEASE_601/final)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin`

When building with the same version of vanilla clang compiler, the issue does not appear. I've also tried clang-7.0 and gcc 8.3.0.

Attached below is the core dump file.
[coredump.zip](https://github.com/CESNET/libyang/files/3425626/coredump.zip)
